### PR TITLE
feat: make staleness check more robust

### LINKED ISSUE #71: Can't use proper-lockfile while debugging memory leaks
I'm trying to debug memory leaks which involves generating heap dumps.  This tends to peg the CPU - when it's done, the process explodes with:

```
/path/to/node_modules/proper-lockfile/lib/lockfile.js:181
        onCompromised: (err) => { throw err; },
                                  ^

Error: Unable to update lock within the stale threshold
    at options.fs.utimes (/path/to/node_modules/proper-lockfile/lib/lockfile.js:109:21)
    at FSReqWrap.oncomplete (fs.js:141:20)
```

### PR DESCRIPTION
**Pull request by hugomrdias**
now we keep track of the mtime updated by the current process, and if we lose some cycles in the update process but recover and the mtime is still ours we do not mark the lock as compromised

closes #71 

ref:
https://github.com/ipfs/js-ipfs-repo/issues/188#issuecomment-468682971

### CODE REVIEW

**Comment by satazor** [comment]
@hugomrdias CI is failing for some reason, can you also check that?

File: lib/lockfile.js [hunk 1]
```diff
@@ -97,43 +97,52 @@ function updateLock(file, options) {
...
-            // Verify if we are within the stale threshold
-            if (lock.lastUpdate <= Date.now() - options.stale && lock.lastUpdate > Date.now() - (options.stale * 2)) {
-                const err = Object.assign(
-                    new Error(lock.updateError || 'Unable to update lock within the stale threshold'),
-                    { code: 'ECOMPROMISED' }
-                );
+        // Check if mtime is still ours
+        options.fs.stat(getLockFile(file, options), (err, stat) => {
+            // Check mtime ownership
+            const isMtimeOurs = options.checkMtimeOwnership && lock.stat.mtimeMs === (stat ? stat.mtimeMs : null);
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply lib/lockfile.js]: this is just for testing and it is enabled by default

File: lib/lockfile.js [hunk 2]
```diff
@@ -97,43 +97,52 @@ function updateLock(file, options) {
...
-                return;
-            }
-
-            // Verify if we are within the stale threshold
-            if (lock.lastUpdate <= Date.now() - options.stale && lock.lastUpdate > Date.now() - (options.stale * 2)) {
-                const err = Object.assign(
-                    new Error(lock.updateError || 'Unable to update lock within the stale threshold'),
-                    { code: 'ECOMPROMISED' }
-                );
+        // Check if mtime is still ours
```
[the remark this answers is not in the document]
Comment (hugomrdias) [reply lib/lockfile.js]: done

File: lib/lockfile.js [hunk 3]
```diff
@@ -209,6 +229,8 @@ function lock(file, options, callback) {
 
                 // We now own the lock
                 const lock = locks[file] = {
+                    stat,
```
Comment (satazor) [inline lib/lockfile.js]: Can we remove this as it’s not needed?

File: lib/lockfile.js [hunk 4]
```diff
@@ -93,33 +93,19 @@ function updateLock(file, options) {
 
     lock.updateDelay = lock.updateDelay || options.update;
     lock.updateTimeout = setTimeout(() => {
-        const mtime = Date.now() / 1000;
+        const mtime = new Date();
```
Comment (satazor) [inline lib/lockfile.js]: Can this be moved near to the .utimes?

File: lib/lockfile.js [hunk 5]
```diff
@@ -93,33 +93,19 @@ function updateLock(file, options) {
...
-                );
-
-                return setLockAsCompromised(file, lock, err);
-            }
-
-            // If the file is older than (stale * 2), we assume the clock is moved manually,
-            // which we consider a valid case
+        // Check if mtime is still ours if it is we can still recover from a system sleep or a busy event loop
+        options.fs.stat(getLockFile(file, options), (err, stat) => {
+            const isMtimeOurs = options.checkMtimeOwnership && lock.mtime.getTime() === (stat ? stat.mtime.getTime() : null);
```
Comment (satazor) [inline lib/lockfile.js]: Can we remove this option?

**Comment by satazor** [comment]
Feel free to merge! Don’t forget to add the (possible) breaking change in the commit message.

**Comment by satazor** [comment]
@hugomrdias do you mind doing `git pull && npm install`? I've resolved a conflict against master that might need the `package-lock.json` to be updated.

File: lib/lockfile.js [hunk 6]
```diff
@@ -93,33 +93,17 @@ function updateLock(file, options) {
...
-                );
-
-                return setLockAsCompromised(file, lock, err);
-            }
-
-            // If the file is older than (stale * 2), we assume the clock is moved manually,
-            // which we consider a valid case
+        // Check if mtime is still ours if it is we can still recover from a system sleep or a busy event loop
+        options.fs.stat(getLockFile(file, options), (err, stat) => {
+            const isMtimeOurs = lock.mtime.getTime() === (stat ? stat.mtime.getTime() : null);
```
Comment (alanshaw) [inline lib/lockfile.js]: Nit: personally I'd move this after the error handling block so you don't need to use a ternary 😄

File: lib/lockfile.js [hunk 7]
```diff
@@ -129,11 +113,46 @@ function updateLock(file, options) {
...
+                }
+
+                // If it failed to update the lockfile, keep trying unless
+                // the lockfile was deleted or we are over the threshold
+                if (err) {
+                    if (err.code === 'ENOENT' || isOverThreshold) {
+                        return setLockAsCompromised(file, lock, Object.assign(err, { code: 'ECOMPROMISED' }));
+                    }
+
+                    lock.updateError = err;
```
Comment (alanshaw) [inline lib/lockfile.js]: These are assigned to but never used...

**Comment by alanshaw** [comment]
Ok this is working great for me! 👌

**Pull request merged**